### PR TITLE
Tait must be constrained if in sig

### LINKED ISSUE #117861: TAIT decision on "may define implies must define"
Nominating for T-lang to decide whether to adopt the following restriction for the stabilization of TAIT:

> At least until the new trait solver is stabilized, any item that is allowed to define the hidden type of some opaque type *must* define the hidden type of that opaque type.

If adopted, we would call this the "may define implies must define" restriction.

**For the purposes of this question, please set aside how we know whether an item is allowed to define the hidden type; this question is invariant to the details of that.**

This restriction leaves space for the new trait solver.  That solver will consider more uses of the opaque type to be defining and it will implement lazy normalization.  Given this, items that may define the hidden type but do not present the potential for differences between the old and new solver that we would prefer to avoid so as to ease the stabilization of the new trait solver.

With this restriction, TAIT becomes roughly equivalent to RPIT with respect to many of the relevant challenges.

The effect of this restriction would be to reject the following code:

```rust
#![feature(type_alias_impl_trait)]

type Tait = impl Sized;
fn define() -> Tait {} // Defines the hidden type.
fn passthrough(x: Tait) -> Tait { // Allowed to define but does not.
    // ~^ ERROR Any item that may define a hidden type must define that hidden type.
    x
}
```

Please set aside how we know that `define` and `passthrough` are both allowed to define the hidden type of the opaque.  There are many ways that could be notated, and that question is orthogonal to this restriction.

Whether or not this restriction could be lifted later is a question of what is possible under the new trait solver and of what other rules we might decide to adopt that may make this more or less possible.  Adopting this restriction now does not in and of itself close the door to possibly permitting this later.

This rule was described and proposed in the 2023-11-08 T-lang Mini-TAIT design meeting ([minutes](https://hackmd.io/CDj92gZdTzilDWORdKhLQA)): rust-lang/lang-team#233.

## Related

- https://github.com/rust-lang/trait-system-refactor-initiative/issues/32
- https://github.com/rust-lang/trait-system-refactor-initiative/issues/40
- https://github.com/rust-lang/trait-system-refactor-initiative/issues/69
- https://github.com/rust-lang/trait-system-refactor-initiative/issues/74

### PR DESCRIPTION
**Pull request by oli-obk**
r? @compiler-errors

kind of reverts https://github.com/rust-lang/rust/pull/62423, but that PR only removed cycles in cases that we want to forbid now anyway.

see https://rust-lang.zulipchat.com/#narrow/stream/315482-t-compiler.2Fetc.2Fopaque-types/topic/lcnr.20oli.20meeting/near/370712246 for related discussion and motivating example

a TAIT showing up in a signature doesn't just mean it can register a hidden type, but that it must.

This is the [design the types team decided upon](https://hackmd.io/QOsEaEJtQK-XDS_xN4UyQA#Proposal-preferred) for the following reasons

* avoids a hypothetical situation where getting smarter in trait solving could cause new cycle errors or new inference errors to show up, where today something is treated as opaque.
* avoids having the situation where a (minor?) change to a function body causes an error, because its TAIT usage suddenly isn't "opaque enough" anymore.
* avoids having to explain why in some cases you need to put a Tait into a tiny module together with its defining usages. Now you basically gotta always do this, the moment a Tait is in a signature that isn't in the defining scope.
* avoids false-cycle errors
* anything diverging from this pattern needs to either
    * move the TAIT declaration and defining function into their own helper sub module 
    * use the attributes we'll provide in the future to explicitly opt in or out of being in the defining scope

fixes #117861

reverts #125362 cc @Nilstrieb @joboet

### CODE REVIEW

File: tests/ui/type-alias-impl-trait/issue-63355.rs [hunk 1]
```diff
@@ -1,5 +1,4 @@
 #![feature(type_alias_impl_trait)]
-// check-pass
```
Comment (oli-obk) [inline tests/ui/type-alias-impl-trait/issue-63355.rs:2]: This is the only test for which there was no workaround after this change.

File: tests/ui/type-alias-impl-trait/type-alias-impl-trait-fns.rs [hunk 2]
```diff
@@ -1,5 +1,3 @@
-// check-pass
-
```
Comment (oli-obk) [inline tests/ui/type-alias-impl-trait/type-alias-impl-trait-fns.rs:2]: This PR "reopens" https://github.com/rust-lang/rust/issues/61863

the test doesn't really add too much new test coverage compared with other tests we already have.

**Comment by compiler-errors** [comment]
I need to take another look at this, but blocking until it's been approved from the T-lang side.

@rustbot blocked

**Comment by oli-obk** [comment]
Nominating for lang team discussion. The explanation is entirely in the main post of this PR.

The question we want T-lang to answer is:

Is it ok from your side that we disallow

```rust
type Tait = impl Sized;
fn foo(x: Tait) -> Tait { x }
fn bar() -> Tait { }
```

but allow the same code if the `foo` function is removed (or moved to a parent or sibling module)

**Comment by traviscross** [comment]
This is also nominated as #117861.  Also see #117865.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

Let's unnominate this in favor of #117861 which gets at the same question.

**Comment by Dylan-DPC** [comment]
@oli-obk what's the status of this? it was blocked as it was nominated but then unnominated for a different pr, so i'm assuming that it is either superseded or blocked by it

**Comment by oli-obk** [comment]
I need to revert https://github.com/rust-lang/rust/pull/125362, because these are annoying to support until we have `#[defines]` attributes

**Comment by lcnr** [comment]
given that this requirement only holds for TAIT, is it possible to not split `from_signature` and `in_body` and instead simply require it for all TAIT?

**Comment by oli-obk** [comment]
> given that this requirement only holds for TAIT, is it possible to not split `from_signature` and `in_body` and instead simply require it for all TAIT?



> given that this requirement only holds for TAIT, is it possible to not split `from_signature` and `in_body` and instead simply require it for all TAIT?

yes, this makes some patterns impossible until we get `#[defines]`, but that seems ok

File: compiler/rustc_hir_analysis/src/collect/type_of/opaque.rs [hunk 3]
```diff
@@ -259,6 +260,16 @@ impl TaitConstraintLocator<'_> {
...
+            if opaque_types_defined_by.contains(&self.def_id) {
+                self.tcx.dcx().emit_err(TaitForwardCompat2 {
+                    span: self
+                        .tcx
+                        .def_ident_span(item_def_id)
+                        .unwrap_or_else(|| self.tcx.def_span(item_def_id)),
+                    opaque_type_span: self.tcx.def_span(self.def_id),
+                    opaque_type: self.tcx.def_path_str(self.def_id),
+                });
+            }
```
Comment (oli-obk) [inline compiler/rustc_hir_analysis/src/collect/type_of/opaque.rs:272]: This is now the only relevant diff in the compiler. Everything else is expected fallout

File: tests/ui/type-alias-impl-trait/issue-109054.rs [hunk 4]
```diff
@@ -11,6 +11,7 @@ impl std::ops::Deref for CallMe {
     type Target = FnType;
 
     fn deref(&self) -> &Self::Target {
+        //~^ ERROR: item does not constrain `ReturnType
         fn inner(val: &u32) -> ReturnType {
```
Comment (lcnr) [inline tests/ui/type-alias-impl-trait/issue-109054.rs:15]: that one is an interesting pattern 🤔 how would we support that one with a `defines` attribute?

**Comment by lcnr** [comment]
@bors r+ rollup=never

going to be prone to merge conflicts given the amount of tests this changes

cc @joboet this reverts #125362, as it relies on non-defining uses of opaque types in the defining scope, to unblock a necessary change for TAIT. Once we've introduce an attribute to explicitly state which functions define opaque types, this PR can be landed again.

**Comment by lqd** [comment]
Probably noise + we need the fix => @rustbot label: +perf-regression-triaged

**Pull request merged**